### ISSUE TITLE
[runtime env] pip-based runtime env refactor

### ISSUE THREAD

**Issue by Catch-Bull** [ray-project/ray#21952]
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

## Work Items

1. Introduce Virtualenv, detail:
    - Create vituralenv for jobs or workers
    - check ray version
    - use `pip check` to find python dependencies conflict
    - download dependencies
    - PR: 
        - https://github.com/ray-project/ray/pull/21801/
        - https://github.com/ray-project/ray/pull/22309
2. Use coroutine in runtime env agent, detail: #21950 
3. Introduce `DownloadManager`, detail：
    - limit download peak.
    - singletons, manager all download tasks
4. Use coroutine to setup `py_modules` and `working_dir`, so that they can be executed in parallel with virtualenv creation.
5. Support change pip version
6.  Install conda and pip in `container`
7. Optimize `pip install $HTTP_URL`, detail:
    - we download package from $HTTP_URL
    - and pip install /path/to/this/package
 8. Support cached Virtualenv

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

**Comment by architkulkarni** [comment]
This plan sounds good to me!

**Comment by architkulkarni** [comment]
> 1. Support pip check after pip install in runtime env:
>    
>    * Whether to support users to disable pip check, in practice, many times users don't care if dependencies have any conflict
>    * If we support user disable pip check, what should the interface of runtime env look like? my proposal:
>      Add field `enable_pip_check`, and runtime_env look like this:
>      ```python
>      runtime_env = {"pip": [...], "enable_pip_check": False}
>      ```

Makes sense! 
> 2. Support user to choose pip version:

Makes sense, I also experienced the looping download behavior in the past.

> 3. Set timeout for runtime env creation and support user change it:

Agree this makes sense.  Should this be on a per-field basis (like one timeout for `pip`, one timeout for `working_dir`, etc) or should it be for the entire runtime env?
> 4. Support user uses pip command directly:

Why is this necessary? Can't the options of pip be specified in `requirements.txt`, or per-line like `["requests", "--extra-index-url=url --some-other-option my_pkg"]`?  Or is that not possible for certain options?

----

As you mentioned, there's two questions about where to put all this, one is about the user interface and one is about the protobuf structure.

For the user interface, adding them like `runtime_env = {"pip": [...], "enable_pip_check": False}` seems fine.  One potential downside is that the extra options for `pip` will appear alongside any (future) extra options for `working_dir`, `py_modules` and so on, so the top level of the runtime env dict could get cluttered.

Another approach would be to have a nested structure like `runtime_env = {"pip": {"packages": [...], "enable_pip_check": False, "version": 21.2.3, ...}, "working_dir": ...}`, but we also need to keep backwards compatibility

For the proto structure, I'm not sure whether things should go in `RuntimeEnv` or `RuntimeEnvInfo`.  It seems like it's less confusing to have them in `RuntimeEnv` to match the interface but I'm not confident about this, what are your thoughts?  Actually, can these two be unified somehow?

**Comment by edoakes** [comment]
I would prefer if we could keep all options related to a given field inside of a single value in the dictionary, but agree with @architkulkarni's point about verbosity.

Is it possible we could change these options at the cluster level instead of per-runtime env? That might be a good solution that'd avoid us needing to add a bunch of new options to the runtime_env schema.

**Comment by Catch-Bull** [comment]
> > 4\. Support user uses pip command directly:
> 
> Why is this necessary? Can't the options of pip be specified in `requirements.txt`, or per-line like `["requests", "--extra-index-url=url --some-other-option my_pkg"]`? Or is that not possible for certain options?

The reason for mentioning `Feature 4` is that we already have this usage internally. It is undeniable that this usage is very tricky, and it will bring many problems...So, I agree with you, we do not support it in runtime env.



> Another approach would be to have a nested structure like `runtime_env = {"pip": {"packages": [...], "enable_pip_check": False, "version": 21.2.3, ...}, "working_dir": ...}`, but it's a little more verbose in the common use case with no extra options. We also need to keep backwards compatibility. Maybe we can change this for Ray 2.0 if we decide it's better.

I also support designing a new nested structure to describe pip in runtime env, maybe we can support both usages at the same time, that is, both of the following are legal:

``` python
runtime_env = {"pip": ["PACKAGE_A", "PACKAGE_B", ...]}
runtime_env = {"pip": {"packages": [...], "enable_pip_check": False, "version": 21.2.3, ...}}
```
Once we guarantee forward compatibility, we don't need to wait until ray 2.0. The main reason is that this also affects the stability of runtime env. As mentioned earlier, some packages may require a higher version of pip. Once the pip version is incorrect, it may be solved by redeploying the ray cluster.

> For the proto structure, I'm not sure whether things should go in `RuntimeEnv` or `RuntimeEnvInfo`. It seems like it's less confusing to have them in `RuntimeEnv` to match the interface but I'm not confident about this, what are your thoughts? Actually, can these two be unified somehow?

I think `setup_runtime_env_timeout_seconds` should be in `RuntimeEnvInfo`, cause it should not participate in the hash value calculation of runtime_env, for the same reason, `enable_pip_check` and `pip_version` should be in `RuntimeEnv`. But modifying `RuntimeEnvInfo` will bring a lot of modifications, I will open a new PR about `setup_runtime_env_timeout_seconds` tomorrow, you can see it in this PR.

**Comment by Catch-Bull** [comment]
About Timeout PR, I will finish it on Monday. and which of the two interfaces do you guys prefer?
```python
@ray.remote(runtime_env = {"pip": ..., "working_dir":..., "setup_runtime_env_timeout_seconds": ...})
def f():
    ...
```
or 
```python
@ray.remote(runtime_env = {"pip": ..., "working_dir":...}, setup_runtime_env_timeout_seconds=...)
def f():
    ...
```

**Comment by architkulkarni** [comment]
I agree that we should have `pip_check`, `pip_version` and `setup_runtime_env_timeout_seconds` per runtime-env rather than cluster env, for the reason you mentioned.  The timeout_seconds especially makes sense as a runtime_env option because certain environments are much slower to install than others.

Since the timeout shouldn't be part of the hash like you said, I think the second interface you proposed makes more sense, but I don't have a strong opinion about it.

**Comment by SongGuyang** [comment]
I really like the second one. But we already have a `eager_install` option in the runtime env dict. I think `eager_install` and `setup_runtime_env_timeout_seconds` should be placed in the same level. @edoakes Do we plan to move it out in future(version 2.0)? It's ok for me to keep it inside `runtime_env` and follow the first interface.

**Comment by architkulkarni** [comment]
I agree about `eager_install`, I think it would make sense to move it out for Ray 2.0.

**Comment by edoakes** [comment]
Ok, seems like we have two separate question here:
1. What to do about the advanced `pip` options.
2. What to do about general runtime_env configurations like `eager_install` and `setup_timeout_s`.

For question (1), I think I have a slight preference for overloading the schema and allowing users to either pass a simple list or a more complex dict that contains more options.

For question (2), I don't want to clutter the top-level `@ray.remote` API with a bunch of runtime_env-related options. What if we instead support a `RuntimeEnvConfig` that contains these options such as the timeout, eager_install, etc.? For example:

```python
@ray.remote(runtime_env=..., runtime_env_config=RuntimeEnvConfig(timeout_s=20, eager_install=False)) # Also support dict.
def f():
    pass
```

Btw, I'm not convinced that we need to support specifying the pip version per-runtime_env. Why can't this be set at the cluster level? It seems unmaintainable for a user to have a bunch of different pip versions running in the same cluster anyways (I'm surprised you allow this internally).

**Comment by edoakes** [comment]
cc @ericl @pcmoritz any thoughts on the above API proposals?

**Comment by SongGuyang** [comment]
> ```python
> @ray.remote(runtime_env=..., runtime_env_config=RuntimeEnvConfig(timeout_s=20, eager_install=False)) # Also support dict.
> def f():
>     pass
> ```

This API looks good for me. Not sure if `RuntimeEnvConfig` is the best name.

**Comment by SongGuyang** [comment]
> Btw, I'm not convinced that we need to support specifying the pip version per-runtime_env. Why can't this be set at the cluster level? It seems unmaintainable for a user to have a bunch of different pip versions running in the same cluster anyways (I'm surprised you allow this internally).

Actually, we prefer to make the package check more strict and standard. We should try to use the higher pip version. But in Ant internal, persons are using different pip and python version in their developing environment. When they board ray cluster, the environment setup fails with built-in pip in some cases. They will say that: "It works in my desktop, why fails in ray cluster?"
So, we allow users walking around some pip issues by a job level config `pip_version`.

**Comment by SongGuyang** [comment]
In addition, some pip issues are brought by some third-party libraries. It's hard to ask users to address that.

**Comment by architkulkarni** [comment]
The use case makes sense, and I think RuntimeEnvConfig is a reasonable name. I'm happy with moving forward with this.

**Comment by edoakes** [comment]
@SongGuyang @Catch-Bull actually, to make this even less intrusive, what if we put these options nested inside of the `runtime_env` field? That would prevent us from needing to update schemas all over the place to support it. For example:

```python
@ray.remote(runtime_env=RuntimeEnv(..., config=RuntimeEnvConfig(setup_timeout_s=10))
def f(): pass

@ray.remote(runtime_env={..., "config": {"setup_timeout_s": 10}})
def f(): pass
```

I would be happy to go forward with this proposal, it is the least intrusive I can think of (I don't think most users will ever think about these advanced options).

**Comment by SongGuyang** [comment]
@edoakes sounds good. We should also move `eager_install` into `RuntimeEnvConfig`, right? Do we need to make `eager_install` backward compatible?

**Comment by edoakes** [comment]
@SongGuyang yes, we should maintain backwards compatibility here. For now, we can leave a top-level `eager_install` alias in `RuntimeEnv` and just set the `config` option if users use it.
